plugins/outputs/otlp/otlp.go

```diff
@@ -0,0 +1,379 @@
+		},
+		InstrumentationLibraryMetrics: []*otlpmetricpb.InstrumentationLibraryMetrics{
+			{
+				InstrumentationLibrary: &otlpcommonpb.InstrumentationLibrary{
+					Name:    instrumentationLibraryName,
+					Version: internal.Version(),
+				},
+				Metrics: []*otlpmetricpb.Metric{metric},
+			},
+		},
+	}, metric
+}
+
+func intGauge(labels []*otlpcommonpb.StringKeyValue, ts int64, value int64) *otlpmetricpb.IntGauge {
+	integer := &otlpmetricpb.IntDataPoint{
+		Labels:       labels,
+		TimeUnixNano: uint64(ts),
+		Value:        value,
+	}
+	return &otlpmetricpb.IntGauge{
+		DataPoints: []*otlpmetricpb.IntDataPoint{integer},
+	}
+}
+
+func doubleGauge(labels []*otlpcommonpb.StringKeyValue, ts int64, value float64) *otlpmetricpb.DoubleGauge {
+	double := &otlpmetricpb.DoubleDataPoint{
+		Labels:       labels,
+		TimeUnixNano: uint64(ts),
+		Value:        value,
+	}
+	return &otlpmetricpb.DoubleGauge{
+		DataPoints: []*otlpmetricpb.DoubleDataPoint{double},
+	}
+}
+
+// Write the metrics to OTLP destination
+func (o *OTLP) Write(metrics []telegraf.Metric) error {
+	batch := sorted(metrics)
+	samples := []*otlpmetricpb.ResourceMetrics{}
+	currentTs := time.Now().UnixNano()
+	for _, m := range batch {
+		for _, f := range m.FieldList() {
+			sample, point := o.protoTimeseries(m, f)
+
+			labels := protoStringLabels(m.TagList())
+			ts := m.Time().UnixNano()
+
+			switch m.Type() {
+			case telegraf.Counter:
+				switch v := f.Value.(type) {
+				case uint64:
+					if v <= uint64(MaxInt) {
+						point.Data = &otlpmetricpb.Metric_IntSum{
+							IntSum: monotonicIntegerPoint(labels, ts, currentTs, int64(v)),
+						}
+					} else {
+						point.Data = &otlpmetricpb.Metric_IntSum{
+							IntSum: monotonicIntegerPoint(labels, ts, currentTs, int64(MaxInt)),
+						}
+					}
+				case int64:
+					point.Data = &otlpmetricpb.Metric_IntSum{
+						IntSum: monotonicIntegerPoint(labels, ts, currentTs, v),
+					}
+				case float64:
+					point.Data = &otlpmetricpb.Metric_DoubleSum{
+						DoubleSum: monotonicDoublePoint(labels, ts, currentTs, v),
+					}
+				case bool:
+					if v {
+						point.Data = &otlpmetricpb.Metric_IntSum{
+							IntSum: monotonicIntegerPoint(labels, ts, currentTs, 1),
+						}
+					} else {
+						point.Data = &otlpmetricpb.Metric_IntSum{
+							IntSum: monotonicIntegerPoint(labels, ts, currentTs, 0),
+						}
+					}
+				case string:
+					o.Log.Error("get type failed: unsupported telegraf value type string")
+					continue
+				default:
+					o.Log.Errorf("get type failed: unsupported telegraf value type %v\n", f.Value)
+					continue
+				}
+			case telegraf.Gauge, telegraf.Untyped:
+				switch v := f.Value.(type) {
+				case uint64:
+					if v <= uint64(MaxInt) {
+						point.Data = &otlpmetricpb.Metric_IntGauge{
+							IntGauge: intGauge(labels, ts, int64(v)),
+						}
+					} else {
+						point.Data = &otlpmetricpb.Metric_IntGauge{
+							IntGauge: intGauge(labels, ts, int64(MaxInt)),
+						}
+					}
```

Comment:
Why do you convert the `uint64` value to int? Anyway this might make more clear that you clip the value:
```suggestion
					x := int64(v)
					if v > uint64(MaxInt) {
						x := MaxInt
					}
					point.Data = &otlpmetricpb.Metric_IntGauge{
						IntGauge: intGauge(labels, ts, x),
					}
```

Reply:
The conversion happens because OTLP only supports `int64`, I've updated the code to something similar to your suggestion.